crates/oxc_sourcemap/src/encode.rs

```diff
@@ -26,37 +28,39 @@ pub fn encode(sourcemap: &SourceMap) -> JSONSourceMap {
 // Here using `serde_json::to_string` to serialization `names/source_contents/sources`.
 // It will escape the string to avoid invalid JSON string.
 pub fn encode_to_string(sourcemap: &SourceMap) -> Result<String> {
-    let mut buf = String::new();
-    buf.push_str("{\"version\":3,");
+    let mut contents = PreAllocatedString::new(
+        10 + sourcemap.names.len() * 2
+            + sourcemap.sources.len() * 2
+            + if let Some(x) = &sourcemap.x_google_ignore_list { x.len() * 2 } else { 0 },
+    );
+    contents.push("{\"version\":3,".into());
     if let Some(file) = sourcemap.get_file() {
-        buf.push_str("\"file\":\"");
-        buf.push_str(file);
-        buf.push_str("\",");
+        contents.push("\"file\":\"".into());
+        contents.push(file.into());
+        contents.push("\",".into());
     }
     if let Some(source_root) = sourcemap.get_source_root() {
-        buf.push_str("\"sourceRoot\":\"");
-        buf.push_str(source_root);
-        buf.push_str("\",");
+        contents.push("\"sourceRoot\":\"".into());
+        contents.push(source_root.into());
+        contents.push("\",".into());
+    }
+    contents.push("\"names\":[".into());
+    for n in &sourcemap.names {
+        contents.push(serde_json::to_string(n.as_ref())?.into());
+        contents.push(",".into());
     }
-    buf.push_str("\"names\":[");
-    let names = sourcemap
-        .names
-        .iter()
-        .map(|x| serde_json::to_string(x.as_ref()))
-        .collect::<std::result::Result<Vec<_>, serde_json::Error>>()
-        .map_err(Error::from)?;
-    buf.push_str(&names.join(","));
-    buf.push_str("],\"sources\":[");
-    let sources = sourcemap
-        .sources
-        .iter()
-        .map(|x| serde_json::to_string(x.as_ref()))
-        .collect::<std::result::Result<Vec<_>, serde_json::Error>>()
-        .map_err(Error::from)?;
-    buf.push_str(&sources.join(","));
+    // Remove the last `,`.
+    contents.pop();
```

Comment:
I think this needs a check that `names` in not empty. If it is empty, then the loop won't execute and `.pop()` will pop off `"names":[`.

```rs
if !sourcemap.names.is_empty() {
    for n in &sourcemap.names {
        contents.push(serde_json::to_string(n.as_ref())?.into());
        contents.push(",".into());
    }
    contents.pop();
}
```